cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java

```diff
@@ -0,0 +1,616 @@
+                                    vertx.cancelTimer(t);
+                                    p.fail(userTaskResult.cause());
+                                }
+                            });
+                        }
+                    } else {
+                        log.error("Cruise Control getting rebalance resource failed", getResult.cause());
+                        vertx.cancelTimer(t);
+                        p.fail(getResult.cause());
+                    }
+                });
+
+            });
+        } else {
+            p.complete(clusterRebalance.getStatus());
+        }
+        return p.future();
+    }
+
+    /**
+     * This method handles the transition from {@code Stopped} state.
+     * If the user set strimzi.io/rebalance=restart annotation, it calls the Cruise Control API for requesting a new rebalance proposal.
+     * If the proposal is just ready, the next state is {@code ProposalReady}.
+     * If the proposal is not ready yet and Cruise Control is still taking care of processing it, the next state is {@code PendingProposal}.
+     * If the user sets any other values for the strimzi.io/rebalance, it is just ignored.
+     *
+     * @param host Cruise Control service to which sending the rebalance proposal request
+     * @param apiClient Cruise Control REST API client instance
+     * @param rebalanceAnnotation The current value for the strimzi.io/rebalance annotation
+     * @param rebalanceOptionsBuilder builder for the Cruise Control REST API client options
+     * @return a Future with the next {@code KafkaClusterRebalanceStatus} bringing the state
+     */
+    private Future<KafkaClusterRebalanceStatus> onStop(String host, CruiseControlApi apiClient,
+                                                       RebalanceAnnotation rebalanceAnnotation,
+                                                       RebalanceOptions.RebalanceOptionsBuilder rebalanceOptionsBuilder) {
+        if (rebalanceAnnotation == RebalanceAnnotation.restart) {
+            return apiClient.rebalance(host, 9090 /* TODO: need public constant from CruiseControl model */, rebalanceOptionsBuilder.build())
+                    .map(response -> {
+                        if (response.getJson() != null) {
+                            return new KafkaClusterRebalanceStatusBuilder()
+                                    .withOptimizationResult(response.getJson().getJsonObject("rebalance").getMap())
+                                    .addNewCondition().withNewType(State.ProposalReady.toString()).endCondition().build();
+                        } else {
+                            return new KafkaClusterRebalanceStatusBuilder()
+                                    .withNewSessionId(response.getUserTaskId())
+                                    .addNewCondition().withNewType(State.PendingProposal.toString()).endCondition().build();
+                        }
+                    });
```

Comment:
If we're not doing to use a transition to the new state to do this, then I guess you can factor out a method to do this and call it from here and `onNew()`.